tests/test_commands/test_utils.py

```diff
@@ -0,0 +1,232 @@
+        DensityOutliersTransform(in_column="target"),
+    ]
+
+    pipeline = Pipeline(transforms=transforms, model=LinearPerSegmentModel(), horizon=14)
+    return pipeline
+
+
+@pytest.mark.parametrize(
+    "num_points, horizon, stride, context_size, expected",
+    (
+        (13, 2, 2, 2, 5),
+        (13, 2, 1, 2, 10),
+        (13, 2, 2, 1, 6),
+        (13, 2, 1, 1, 11),
+        (13, 1, 1, 1, 12),
+        (13, 4, 4, 6, 1),
+        (13, 4, 1, 6, 4),
+        (10, 5, 1, 5, 1),
+        (10, 5, 5, 5, 1),
+    ),
+)
+def test_estimate_n_folds(num_points, horizon, stride, context_size, expected):
+    res = _estimate_n_folds(num_points=num_points, horizon=horizon, stride=stride, context_size=context_size)
+    assert res == expected
+
+
+def test_estimate_n_folds_not_enough_points(num_points=10, horizon=7, stride=1, context_size=5):
+    with pytest.raises(ValueError, match="Not enough data points!"):
+        _ = _estimate_n_folds(num_points=num_points, horizon=horizon, stride=stride, context_size=context_size)
+
+
+def test_estimate_n_folds_forecast_no_ts(pipeline_without_context):
+    with pytest.raises(ValueError, match="There is no ts for forecast method!"):
+        _ = _max_n_folds_forecast(pipeline=pipeline_without_context, ts=None, context_size=1)
+
+
+def test_estimate_n_folds_backtest_no_ts(pipeline_without_context):
+    with pytest.raises(ValueError, match="Parameter `ts` is required when estimating for backtest method"):
+        _ = estimate_max_n_folds(pipeline=pipeline_without_context, method_name="backtest", context_size=1)
+
+
+def test_estimate_n_folds_backtest_intervals_error(pipeline_without_context, example_tsds):
+    with pytest.raises(
+        NotImplementedError, match="Number of folds estimation for backtest with intervals is not implemented!"
+    ):
+        _ = _max_n_folds_backtest(
+            pipeline=pipeline_without_context,
+            ts=example_tsds,
+            forecast_params={"prediction_interval": True},
+            context_size=1,
+        )
+
+
+def test_estimate_max_n_folds_invalid_method_name(pipeline_without_context, example_tsds, method_name="fit"):
+    with pytest.raises(ValueError, match="fit is not a valid method name."):
+        _ = estimate_max_n_folds(
+            pipeline=pipeline_without_context, ts=example_tsds, method_name=method_name, context_size=1
+        )
+
+
+def test_estimate_max_n_folds_empty_ts(pipeline_without_context, empty_ts):
+    with pytest.raises(ValueError, match="Empty ts is passed!"):
+        _ = estimate_max_n_folds(pipeline=pipeline_without_context, ts=empty_ts, method_name="forecast", context_size=1)
+
+
+def test_estimate_max_n_folds_negative_context(pipeline_without_context, example_tsds):
+    with pytest.raises(ValueError, match="Pipeline `context_size` parameter must be positive integer!"):
+        _ = estimate_max_n_folds(
+            pipeline=pipeline_without_context, ts=example_tsds, method_name="forecast", context_size=-1
+        )
+
+
```

Comment:
Have you checked that if `pipeline` already has `ts` inside the `ts` can be not passed into the function? We can also check that if both are present (ts in pipeline and ts passed into the function) the second is used.

Reply:
First point is already tested. Added test for second point

Reply:
Can you show where the first point is tested?